### DISCUSSION TITLE
Compatibility for iOS?

### DISCUSSION THREAD

**Discussion by CodeRunner5235** [ptitSeb/box86#675] [General]
I would love to be able to use something like this to finally be able to do something so extreme it will make every apple lover cry out in rage... run Undertale on an iPhone. In all seriousness though, it would be awesome if you could make it work on iOS. Although I can see that iOS is very different from linux and it would probably require a lot of work. I am honestly just curious if its possible or if ill be stuck with UTM.

**Comment by ptitSeb** [comment]
Box86/box64 are not system emulator, but more "compatibility layer", using the actual system on whta it is running on and "just" translate the CPU, keeping most of the libraries calls as-is.

I'm not familiar with the inners of iOS (is it based on BSD like macOS?), but, to give an example, even Android (that is based on Linux), box86 doesn't run directly and needs a software like "Termux" to have a full Linux system to then run box86.

So, I guess a Termux like software would be needed first.

**Reply by kenny000001** [reply]
You can use iOS with Jailbreak (such as Dopamine Jailbreak),then you can install terminal app for iOS.But for some reason(like your system's version are 16.xx.xx or above),you couldn't directly change the root files,this problem caused by CSV(I think I may remember a wrong word....).(I mean that you can use the user and root account,but you couldn't directly change the files in / ,you also can use with bindfs-mount to change the / files,but it also couldn't change / files as stable.)

**Reply by kenny000001** [reply]
And it's also based on BSD
(macOS and iOS are based on Darwin which open source by apple(in github it called xnu...?))

**Comment by ptitSeb** [comment]
(Converting this ticket to a discussion)

**Comment by CodeRunner5235** [comment]
I believe that iOS is based on BSD. It seams like its a heavily modified version of macOS. As for Termux, it seams like the closest thing for iOS is iSH, which is "A project to get a Linux shell running on iOS, using usermode x86 emulation and syscall translation."

**Comment by duckcantcode** [comment]
Yeah, I'd absolutely like to see box86 and box64 on iOS. This could open up a new world of possibilities for running x86_64 Linux-based Gamemaker and Unity games on iOS.

Yes, [iOS is at least partially based on BSD](https://en.wikipedia.org/wiki/IOS). 

I use iSH a lot but I'm not sure an App Store app (i.e. no JIT) has enough power to run something like box86 at a good speed. To run as a native app, the box86 developer would probably have to figure out how to port the dynarec to iOS. It's certainly possible and probably doable, it'd just take some time. But getting Linux userspace functionality as an iOS app would be difficult.

On the other hand, iSH provides exactly what we need in terms of getting something like Linux's userspace on iOS. Someone could fork it and add non-App Store JIT that only works when activated with AltJIT/SideJIT/TrollJIT/etc. iSH's "JIT" is specialized and [is similar to an interpreted emulator (see the README)](https://github.com/ish-app/ish#a-note-on-the-jit) so that it can be approved for the App Store.

Jailbroken devices expose userspace and this could be done there too if we could figure out how to get box86 to work directly on the iOS kernel/BSD, but that's not as practical since that's way more difficult than sideloading an app.

In the mean time, *does* box86 work on iSH? If so, how?

**Reply by kenny000001** [reply]
Emmm, I remember iSH already a x86(or x64?) simulator for iOS,I don't understand why you install it in iSH?Is it un-useful?

**Reply by duckcantcode** [reply]
> Emmm, I remember iSH already a x86(or x64?) simulator for iOS,I don't understand why you install it in iSH?Is it un-useful?

@kenny000001 It was an example of one way of providing a Linux userspace equivalent on iOS.
